src/v/kafka/server/write_at_offset_stm.cc

```diff
@@ -0,0 +1,270 @@
+        if (last_log_offset != stm_last_offset) {
+            enqueued_promise.set_value();
+            vlog(
+              _log.debug,
+              "Expected last log offset: {} does not match with last stm"
+              "tracked offset: {}",
+              last_log_offset,
+              stm_last_offset);
+            co_return make_error_code(errc::invalid_last_log_offset);
+        }
+        // fill the gap with ghost batches if required
+        if (last_log_offset != kafka::prev_offset(expected_base_offset)) {
+            std::ranges::move(
+              model::make_ghost_batches(
+                kafka::offset_cast(stm_expected_next_offset),
+                model::prev_offset(kafka::offset_cast(expected_base_offset)),
+                model::term_id{0}),
+              std::back_inserter(batches));
+        }
+    } else {
+        if (stm_expected_next_offset != expected_base_offset) {
+            vlog(
+              _log.debug,
+              "Expected write offset: {} does not match with stm tracked "
+              "offset: {}",
+              expected_base_offset,
+              stm_expected_next_offset);
+            enqueued_promise.set_value();
+            co_return make_error_code(errc::invalid_offset);
+        }
+    }
+    _inflight_last_offset = kafka::offset(
+      expected_base_offset() + batch.header().last_offset_delta);
+    batches.push_back(std::move(batch));
+
+    auto stages = _raft->replicate_in_stages(
+      _insync_term,
+      std::move(batches),
+      raft::replicate_options(raft::consistency_level::quorum_ack));
+
+    auto enq = std::move(stages.request_enqueued).finally([u = std::move(u)] {
+    });
+
+    std::move(enq).forward_to(std::move(enqueued_promise));
+
+    co_return co_await std::move(stages.replicate_finished)
+      .handle_exception([this](const std::exception_ptr& e) {
+          _inflight_last_offset.reset();
+          vlog(_log.warn, "Replication failed with exception: {}", e);
+          return result<raft::replicate_result>(
+            make_error_code(errc::replicate_exception));
+      })
+      .then([this](result<raft::replicate_result> result) {
+          if (result.has_error()) {
+              vlog(
+                _log.warn,
+                "Replication failed with an error: {}",
+                result.error());
+
+              if (result.error() != raft::errc::not_leader) {
+                  return _raft->step_down("write_at_offset_replication_failed")
+                    .then([result] mutable { return result; });
+              }
+              _inflight_last_offset.reset();
+          }
+
+          return ssx::now(result);
+      });
+}
+bool write_at_offset_stm::is_offset_translated_batch(
+  const model::record_batch& batch) const {
+    return std::ranges::find(_offset_translated_batches, batch.header().type)
+           != _offset_translated_batches.end();
+}
+
+ss::future<iobuf> write_at_offset_stm::take_snapshot(model::offset) {
```

Comment:
should this be named `take_raft_snapshot`?

Reply:
it is take_raft_snapshot, it is inherited from `stm_base`, i can rename it everywhere in a separate pr